lib/using_restricted_array.rb

```diff
@@ -6,56 +6,125 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n). N is the length of the array.
+# Space complexity: O(1). It always returns one value, so space is constant.
 def length(array)
-  raise NotImplementedError
+  counter = 0
+  
+  while array[counter] != nil
+    counter += 1
+  end
+  
+  return counter
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n). N is the length of the array.
+# Space complexity: O(1). It doesn't store anything in memory, so space is constant.
 def print_array(array)
-  raise NotImplementedError
+  counter = 0
+  
+  while counter != length(array)
```

Comment:
Because you're calling `length(array)` with each iteration, this is causing your method to be an O(n<sup>2</sup>) algorithm.  Think about it this way because `length` is an O(n) method and gets called with each iteration of the while loop.

You can do better if you save `length(array)` into a local variable and use the local variable.  That way it only gets called once.